src/algebra/category/Mon/basic.lean

```diff
@@ -31,9 +31,18 @@ add_decl_doc AddMon
 
 namespace Mon
 
+/-- `monoid_hom` doesn't actually assume associativity. This alias is needed to make the category
+theory machinery work. -/
+@[to_additive "`add_monoid_hom` doesn't actually assume associativity. This alias is needed to make
+the category theory machinery work."]
+abbreviation assoc_monoid_hom (M N : Type*) [monoid M] [monoid N] := monoid_hom M N
+
 @[to_additive]
-instance bundled_hom : bundled_hom @monoid_hom :=
-⟨@monoid_hom.to_fun, @monoid_hom.id, @monoid_hom.comp, @monoid_hom.coe_inj⟩
+instance bundled_hom : bundled_hom assoc_monoid_hom :=
+⟨λ M N [monoid M] [monoid N], by exactI @monoid_hom.to_fun M N _ _,
+ λ M [monoid M], by exactI @monoid_hom.id M _,
+ λ M N P [monoid M] [monoid N] [monoid P], by exactI @monoid_hom.comp M N P _ _ _,
+ λ M N [monoid M] [monoid N], by exactI @monoid_hom.coe_inj M N _ _⟩
```

Comment:
Ah! Once you've done this, you now have the further flexibility to relax the requirements on `monoid_hom` all the way down to `has_one` and `has_mul` (rather than `mul_one_class`).

That would be better, right?

Reply:
Hmm, I've tried this, and while it works in `src/algebra/group/hom.lean`, there are lots of mysterious failures of type inference in later files. :-(

Reply:
I pushed what I did to a branch `non_assoc_monoids_2` if you want to have a look. The diff is at https://github.com/leanprover-community/mathlib/compare/non_assoc_monoids...non_assoc_monoids_2.

Reply:
> Ah! Once you've done this, you now have the further flexibility to relax the requirements on `monoid_hom` all the way down to `has_one` and `has_mul` (rather than `mul_one_class`).
> 
> That would be better, right?

I think that in the past 18 months or so, this has been attempted by Yury, Eric, me, and possibly others. I have no idea why so many problems show up, but I've basically resigned myself to the fact that `monoid_hom` will assume that source and target are `monoid`s.

Reply:
I don't want to open that can of worms in this PR, but I'd encourage you to try it if this one gets merged.